net_connection/src/protocol_wrapper.rs

```diff
@@ -0,0 +1,433 @@
+                    config: json["config"].as_str().unwrap().to_string(),
+                });
+            } else if method == "connect" {
+                assert!(json["address"].is_string());
+                return ProtocolWrapper::Connect(ConnectData {
+                    address: json["address"].as_str().unwrap().to_string(),
+                });
+            } else if method == "peerConnected" {
+                assert!(json["id"].is_string());
+                return ProtocolWrapper::PeerConnected(PeerConnectData {
+                    id: json["id"].as_str().unwrap().to_string(),
+                });
+            } else if method == "send" {
+                assert!(json["_id"].is_string());
+                assert!(json["toAddress"].is_string());
+                return ProtocolWrapper::SendMessage(SendData {
+                    msg_id: json["_id"].as_str().unwrap().to_string(),
+                    to_address: json["toAddress"].as_str().unwrap().to_string(),
+                    data: json["data"].clone(),
+                });
+            } else if method == "handleSend" {
+                assert!(json["_id"].is_string());
+                assert!(json["toAddress"].is_string());
+                assert!(json["fromAddress"].is_string());
+                return ProtocolWrapper::HandleSend(HandleSendData {
+                    msg_id: json["_id"].as_str().unwrap().to_string(),
+                    to_address: json["toAddress"].as_str().unwrap().to_string(),
+                    from_address: json["fromAddress"].as_str().unwrap().to_string(),
+                    data: json["data"].clone(),
+                });
+            } else if method == "handleSendResult" {
+                assert!(json["_id"].is_string());
+                return ProtocolWrapper::HandleSendResult(SendResultData {
+                    msg_id: json["_id"].as_str().unwrap().to_string(),
+                    data: json["data"].clone(),
+                });
+            } else if method == "sendResult" {
+                assert!(json["_id"].is_string());
+                return ProtocolWrapper::SendResult(SendResultData {
+                    msg_id: json["_id"].as_str().unwrap().to_string(),
+                    data: json["data"].clone(),
+                });
+            }
+        }
+
+        ProtocolWrapper::RawProtocol(p.clone())
+    }
+}
+
+impl From<Protocol> for ProtocolWrapper {
+    fn from(p: Protocol) -> Self {
+        ProtocolWrapper::from(&p)
+    }
+}
+
+/// downgrade a ProtocolWrapper instance back into a Protocol message
+impl<'a> From<&'a ProtocolWrapper> for Protocol {
+    // david.b (neonphog) - tarpaulin doesn't cover macros... skipping this fn
+    #[cfg_attr(tarpaulin, skip)]
+    fn from(w: &'a ProtocolWrapper) -> Self {
+        match w {
+            ProtocolWrapper::RawProtocol(p) => p.clone(),
+            ProtocolWrapper::RequestState => Protocol::Json(
+                json!({
```

Comment:
this approach is fragile as the json structure must be manually maintained across everywhere that handles it

a more robust technique is to define a struct/enum and derive `DefaultJson`

something like this:

```rust
#[derive(Debug, Serialize, Deserialize, DefaultJson)]
pub struct RequestStateData{ method: Method }

Protocol::Json(RequestStateData{ method: Method::RequestState }.into())
```

then later:

```rust
RequestStateData::try_from(json)?
```

Reply:
:+1: updated to try_from